File: fretecalc/app/src/main/java/com/example/fretecalc/views/SearchActivity.java
package com.example.fretecalc.views;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.fretecalc.R;
import com.example.fretecalc.viewmodels.SearchActivityViewModel;
import com.example.fretecalc.views.adapters.RecyclerSearchAdapter;
import com.example.fretecalc.views.interfaces.PredictionListener;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.libraries.places.api.Places;
import com.google.android.libraries.places.api.model.AutocompletePrediction;
import com.google.android.libraries.places.api.model.Place;

import java.util.Arrays;
import java.util.List;

import static com.example.fretecalc.utils.ApisConstants.GOOGLE_API_KEY;
import static com.example.fretecalc.views.RequestActivity.DESTINATION_PLACE_CODE;
import static com.example.fretecalc.views.RequestActivity.ORIGIN_PLACE_CODE;

public class SearchActivity extends AppCompatActivity implements PredictionListener {
    private static final int LOCATION_REQUEST_CODE = 1;
    public static final String DESCRIPTION_KEY = "description";
    public static final String LATITUDE_KEY = "latitude";
    public static final String LONGITUDE_KEY = "longitude";
    private EditText searchPlace;
    private TextView textLocation;
    private RecyclerView recyclerPrediction;
    private RecyclerSearchAdapter adapter;
    private FusedLocationProviderClient fusedLocationClient;
    private SearchActivityViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);
        Toolbar toolbar = findViewById(R.id.toolbar_search);
        setSupportActionBar(toolbar);

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayShowTitleEnabled(false);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        initViews();

        searchPlace.requestFocus();
        searchPlace.addTextChangedListener(filterTextWatcher);

        textLocation.setOnClickListener(view -> {
            getLocationData();
        });

        viewModel.getPredictionList().observe(this, autocompletePredictions -> {
            adapter.updatePredictions(autocompletePredictions);
        });

        displayError();
    }

    private void initViews() {
        searchPlace = findViewById(R.id.edit_search_place);
        textLocation = findViewById(R.id.text_search_location);
        recyclerPrediction = findViewById(R.id.recycler_search_autocomplete);
        viewModel = ViewModelProviders.of(this).get(SearchActivityViewModel.class);
        adapter = new RecyclerSearchAdapter(this, this);
        recyclerPrediction.setLayoutManager(new LinearLayoutManager(this));
        recyclerPrediction.addItemDecoration(new DividerItemDecoration(recyclerPrediction.getContext(),
                DividerItemDecoration.VERTICAL));
        recyclerPrediction.setAdapter(adapter);
        adapter.notifyDataSetChanged();
    }

    @Override
    public boolean onSupportNavigateUp() {
        setResult(RESULT_CANCELED);
        finish();
        return true;
    }

    private TextWatcher filterTextWatcher = new TextWatcher() {
        public void afterTextChanged(Editable s) {
            if (!s.toString().equals("")) {
                viewModel.getPredictions(s);
                if (recyclerPrediction.getVisibility() == View.GONE) {
                    recyclerPrediction.setVisibility(View.VISIBLE);
                }
            } else {
                if (recyclerPrediction.getVisibility() == View.VISIBLE) {
                    recyclerPrediction.setVisibility(View.GONE);
                }
            }
        }

        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        }

        public void onTextChanged(CharSequence s, int start, int before, int count) {
        }
    };

    public void onLocationClick(String description, Double latitude, Double longitude) {
        Intent intent = new Intent();
        if (ORIGIN_PLACE_CODE == 100) {
            intent.putExtra(DESCRIPTION_KEY, description);
            intent.putExtra(LATITUDE_KEY, latitude);
            intent.putExtra(LONGITUDE_KEY, longitude);
        } else if (DESTINATION_PLACE_CODE == 200) {
            intent.putExtra(DESCRIPTION_KEY, description);
            intent.putExtra(LATITUDE_KEY, latitude);
            intent.putExtra(LONGITUDE_KEY, longitude);
        }
        setResult(RESULT_OK, intent);
        finish();
    }

    private void getLocationData() {
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this);
        int permissionLocation = ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION);

        if (permissionLocation == PackageManager.PERMISSION_GRANTED) {
            fusedLocationClient.getLastLocation()
                    .addOnSuccessListener(this, location -> {
                        if (location != null) {
                            onLocationClick("Sua localização", location.getLatitude(), location.getLongitude());
                        }
                    });
        } else {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, LOCATION_REQUEST_CODE);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {

        switch (requestCode) {
            case LOCATION_REQUEST_CODE:
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    fusedLocationClient.getLastLocation().addOnSuccessListener(this, location -> {
                        if (location != null) {
                            onLocationClick("Sua localização", location.getLatitude(), location.getLongitude());
                        }
                    });
                } else {
                    Toast.makeText(this, "Permissão negada, escolha manualmente uma cidade", Toast.LENGTH_SHORT).show();
                }
                break;
            default:
                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
                break;
        }
    }

    @Override
    public void onPredictionClick(AutocompletePrediction prediction) {
        String placeId = prediction.getPlaceId();

        viewModel.getPlaceById(placeId);

        viewModel.getPlace().observe(this, place -> {
            Intent intent = new Intent();
            if (ORIGIN_PLACE_CODE == 100) {
                intent.putExtra(DESCRIPTION_KEY, place.getName());
                intent.putExtra(LATITUDE_KEY, place.getLatLng().latitude);
                intent.putExtra(LONGITUDE_KEY, place.getLatLng().longitude);
            } else if (DESTINATION_PLACE_CODE == 200) {
                intent.putExtra(DESCRIPTION_KEY, place.getName());
                intent.putExtra(LATITUDE_KEY, place.getLatLng().latitude);
                intent.putExtra(LONGITUDE_KEY, place.getLatLng().longitude);
            }
            setResult(RESULT_OK, intent);
            finish();
        });
    }

    private void displayError() {
        viewModel.getError().observe(this, s -> {
            Toast.makeText(this, s, Toast.LENGTH_SHORT).show();
        });
    }
}

File: FreightPad/app/src/main/java/dev/khalil/freightpad/ui/activity/SearchActivity.kt
package dev.khalil.freightpad.ui.activity

import android.Manifest.permission.ACCESS_COARSE_LOCATION
import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.location.Location
import android.net.Uri
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.os.Bundle
import android.provider.Settings
import android.text.Editable
import android.text.TextWatcher
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.app.ActivityCompat.checkSelfPermission
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.DividerItemDecoration.VERTICAL
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.gms.location.LocationServices
import com.google.android.material.snackbar.Snackbar
import dev.khalil.freightpad.R
import dev.khalil.freightpad.common.ADDRESS_KEY
import dev.khalil.freightpad.common.LAT_KEY
import dev.khalil.freightpad.common.LONG_KEY
import dev.khalil.freightpad.databinding.ActivitySearchBinding
import dev.khalil.freightpad.di.searchModule
import dev.khalil.freightpad.extensions.gone
import dev.khalil.freightpad.extensions.viewModel
import dev.khalil.freightpad.extensions.visible
import dev.khalil.freightpad.model.Place
import dev.khalil.freightpad.ui.adapter.LocationsAdapter
import dev.khalil.freightpad.ui.adapter.holder.LocationClick
import dev.khalil.freightpad.ui.viewModel.SearchActivityViewModel
import org.kodein.di.Kodein
import org.kodein.di.KodeinAware

class SearchActivity : AppCompatActivity(), KodeinAware, LocationClick {

  override val kodein = Kodein.lazy {
    import(searchModule)
  }

  private lateinit var binding: ActivitySearchBinding

  private val locationsAdapter by lazy { LocationsAdapter(arrayListOf(), this) }
  private val fusedLocationClient by lazy { LocationServices.getFusedLocationProviderClient(this) }

  private val searchViewModel: SearchActivityViewModel by viewModel()

  companion object {
    private const val LOCATION_PERMISSION_CODE = 3

    fun createIntent(context: Context): Intent {
      return Intent(context, SearchActivity::class.java)
    }
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    binding = DataBindingUtil.setContentView(this, R.layout.activity_search)

    initView()
    initRecycler()
    initObservers()
    initListeners()
  }

  override fun locationClicked(place: Place) {
    setLocation(place)
  }

  override fun onRequestPermissionsResult(
    requestCode: Int, permissions: Array<out String>,
    grantResults: IntArray) {

    when (requestCode) {
      LOCATION_PERMISSION_CODE -> {
        if (grantResults.isNotEmpty() && grantResults.first() == PERMISSION_GRANTED) {
          getUserLocation()
        } else {
          val showRationale = if (VERSION.SDK_INT >= VERSION_CODES.M) {
            shouldShowRequestPermissionRationale(permissions.first())
          } else {
            false
          }
          if (!showRationale) {
            Snackbar.make(
              findViewById(android.R.id.content),
              R.string.a_search_permission_permanently_denied,
              Snackbar.LENGTH_LONG)
              .setAction(R.string.ok) {
                startActivity(Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                  this.data = Uri.fromParts("package", packageName, null)
                })
              }
              .show()
          } else {
            Toast.makeText(this, R.string.a_search_permission_denied, Toast.LENGTH_LONG).show()
          }
        }
      }
    }
    super.onRequestPermissionsResult(requestCode, permissions, grantResults)
  }

  override fun onDestroy() {
    searchViewModel.onDestroy()
    super.onDestroy()
  }

  private fun initListeners() {
    binding.myLocationText.setOnClickListener {
      if (isLocationPermissionGranted()) {
        getUserLocation()
      } else {
        requestLocationPermission()
      }

    }
    binding.placeEdit.addTextChangedListener(searchTextWatcher())
    binding.errorView.errorButton.setOnClickListener {
      searchViewModel.search(binding.placeEdit.text.toString())
    }
  }

  private fun initView() {
    binding.placeEdit.requestFocus()
  }

  private fun initRecycler() {
    binding.placesRecycler.apply {
      adapter = locationsAdapter
      addItemDecoration(DividerItemDecoration(this.context, VERTICAL))
      layoutManager = LinearLayoutManager(this@SearchActivity, LinearLayoutManager.VERTICAL, false)
    }
  }

  private fun initObservers() {
    searchViewModel.searchResult.observe(this, Observer { locationList ->
      binding.placesRecycler.visible()
      locationsAdapter.updateLocations(locationList)
    })
    searchViewModel.loading.observe(this, Observer { isVisible ->
      if (isVisible) {
        binding.progressBar.visible()
        binding.placesRecycler.gone()
      } else {
        binding.progressBar.gone()
      }
    })
    searchViewModel.error.observe(this, Observer { isError ->
      if (isError) {
        binding.placesRecycler.gone()
        binding.errorView.root.visible()
      } else {
        binding.errorView.root.gone()
      }
    })
  }

  private fun searchTextWatcher(): TextWatcher {
    return object : TextWatcher {
      override fun afterTextChanged(s: Editable?) {
        searchViewModel.search(s.toString())
      }

      override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {

      }

      override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
      }
    }
  }

  private fun isLocationPermissionGranted(): Boolean {
    return checkSelfPermission(this, ACCESS_FINE_LOCATION) == PERMISSION_GRANTED &&
        checkSelfPermission(this, ACCESS_COARSE_LOCATION) == PERMISSION_GRANTED
  }

  private fun requestLocationPermission() {
    val permissions = arrayOf(ACCESS_FINE_LOCATION, ACCESS_COARSE_LOCATION)
    ActivityCompat.requestPermissions(this, permissions, LOCATION_PERMISSION_CODE)
  }

  @SuppressLint("MissingPermission")
  private fun getUserLocation() {
    fusedLocationClient.lastLocation
      .addOnSuccessListener { location: Location ->
        val place = Place(
          getString(R.string.your_location),
          listOf(location.longitude, location.latitude))
        //TODO TEST THIS SHIT
        setLocation(place)
      }
  }

  private fun setLocation(place: Place) {
    val intent = Intent().apply {
      putExtra(ADDRESS_KEY, place.displayName)
      putExtra(LONG_KEY, place.point.first())
      putExtra(LAT_KEY, place.point.last())
    }
    setResult(Activity.RESULT_OK, intent)
    finish()
  }

}
